# AMS 326 Numerical Analysis HW 2

By: Mehadi Chowdhury (115112722)

Link to GitHub where README, PDF, IPYNB and Seperate Python Files will be hosted: https://github.com/EmceeCiao/AMS_326_HW2 

## Background For Problem 2.1 

<div style="text-align: center;">
    <img src="Problem_2.1_Background.png" alt="Image of Problem 2.1 (Kidney Problem)" style="width:1000px; border:2px solid black;"/>
</div> 



## Problem 2.1 - Task 1 

### Problem Description  

Given a kidney equation of $(x^2 + y^2)^2 = x^3 + y^3$ and digging a disc from the kidney with equation 
$(x-0.25)^2 + (y-0.25)^2 = 0.125$  
Write a program to use the rectangle method to compute the area of the remaining kidney to 4 siginficant digits. 

### Algorithm Description 

There are three rectangle methods that can be used, but the most accurate one is the midpoint method,   
which is the one we will be using here! The midpoint method approximates the area underneath the curve      
by having us evaluate the function value at the midpoint of every interval and take the sum.     
The exact equation of the formula is as follows:    

$\int_{a}^{b} f(x)dx \approx \Delta x[f(x_1) + f(x_2) + ..... + f(x_n)]$   

In this specific case however, we have to do a bit more thinking. Considering a disk is being cut out of the kidney   
and that the disk is in the equation form of a circle, we can easily compute it's area as a constant and subtract   
from the approximation we get using the rectangle method on the kidney.   

To do this method on the kidney, we need a clear way to calculate a y-value, in this case this will require us to   
convert the equation into polar like so: 

$(x^2 + y^2)^2 = x^3 + y^3 \rightarrow r^4 = r^3(sin^3\theta + cos^3\theta) \rightarrow r = sin^3\theta + cos^3\theta$ 

With this we can actually perform the rectangle method as we can treat r as our y-value and the bounds for theta or what
is essentially x in this case is from 0 to 2 $\pi$. This has to be slightly adapted though as we are now in the polar plane, 
each "rectangle" instead of being f(x) * $\Delta x$ it's now $1/2 * r^2 * \Delta \theta$ as it's a circular cut. 

This is derived from the fact that a normal circle is $\pi r^2$ which when only taking a fraction of the circle
we get $(\theta /(2 \pi)) *(\pi r^2)$  
Thus simplifying to what we had stated prior! 

Now that we have our formula, all that's left to decide is the number of intervals we want for our accuracy, in our case 
I believe 10,000 intervals will be enough to get me close to an approximation within 4 significant digits. 

#### Psuedocode  

**Rectangle Method**  

disk_area $\leftarrow$ radius^2 * pi      
kidney_eq $\leftarrow$ cos^3(x) + sin^3(x)    
intervals $\leftarrow$ 10,000 

interval_vals $\leftarrow $ [0, 2pi/intervals, 2* 2pi/intervals, .....]     
$\Delta \leftarrow 2pi/intervals$  

from i in range (0, intervals)    
&emsp; midpoint $\leftarrow$ (interval_vals[i] + interval_vals[i+1])/2  
&emsp; area += (kidney_eq(midpoint))^2 * $\Delta$ * 0.5 

return area - disk_area 


#### Code

In [ ]:
import numpy as np 

def disk_area(r_squared):  
    # We are given r^2 in the formuala for the disk so let the user pass it in 
    return r_squared * np.pi 

def kidney_equation(x):  
    # Equation of Kidney after translating to Polar, x is the angle plugged in!
    return np.cos(x) ** 3 + np.sin(x) ** 3 

def rectangle_method_polar(kidney_eq, intervals, dug_area):      
    """ 
    Function For Midpoint Method on Polar Equation with cut out area! 

    Input: 
    - kidney_eq: equation for the provided kidney (Could be generalized to function of curve) 
    - intervals: Whole number of intervals to be used in calculating 
    - dug_area: Area of disk that's been cut out of the kidney 

    Output: 
    - result: The approximation of the area under the curve after subtracting out the dug area 
              using rectangle method
    """ 
    interval_vals = np.linspace(0, 2*np.pi, intervals+1) 
    step = (2*np.pi)/intervals  
    area = 0
    for i in range(len(interval_vals)-1):     
        midpoint = (interval_vals[i] + interval_vals[i+1])/2
        area += 0.5 * step * (kidney_eq(midpoint) ** 2)   
    return area-dug_area

dug_area = disk_area(0.125) 

ans = rectangle_method_polar(kidney_equation, 10000, dug_area) 
print(f"Approximated Area: {ans}") 

Approximated Area: 1.5707963267949003


### Results 

The results of my program were the following: 

Area of Kidney with Cut Disk $\approx$ 1.570796 

### Performance 

In terms of the performance of this rectangular method polar, the time complexity would be O(n) where n   
represents the number of intervals we decided to use as that's the number of operations performed.    
The space complexity of the program is also O(n) as we have to store all of these intervals. However,  
computationally, this is easier than computing the integral for the area and serves as a good estimate. 


## Problem 2.1 - Task 2

### Problem Description  

Given a kidney equation of $(x^2 + y^2)^2 = x^3 + y^3$ and digging a disc from the kidney with equation 
$(x-0.25)^2 + (y-0.25)^2 = 0.125$  
Write a program to use the trapezoidal method to compute the area of the remaining kidney to 4 siginficant digits. 

### Algorithm Description 

The trapezoidal method approximates the area underneath the curve      
by having us evaluate the function value at the left and right endpoints of every interval, divide it by 2 and multiply it by   
the step as the formula for a trapezoid is $(a+b)/2 * h$. After summing all these values together we get the approximate area   
under the curve.   

The exact equation of the formula is as follows:    

$\int_{a}^{b} f(x)dx \approx \Delta x/2 [f(x_0) + 2f(x_2) + 2f(x_3) + ..... + f(x_n)]$   

In this specific case however, we have to do a bit more thinking just like for task 1! We will have to cut out 
the constant area of the disk which can be calculated easily as it's in the form of a circle. 

Again to use the trapezoidal method on the kidney, we need a clear way to calculate a y-value, in this case this will require us to   
convert the equation into polar like so: 

$(x^2 + y^2)^2 = x^3 + y^3 \rightarrow r^4 = r^3(sin^3\theta + cos^3\theta) \rightarrow r = sin^3\theta + cos^3\theta$ 

With this we can perform our trapezoidal method of integration, again with the caveat that we will have to do r^2   
due to us working in polar coordinates and once again for accuracy I will be using 10,000 intervals.   

#### Psuedocode 

**Trapezoidal Method**  

disk_area $\leftarrow$ radius^2 * pi      
kidney_eq $\leftarrow$ cos^3(x) + sin^3(x)    
intervals $\leftarrow$ 10,000 

interval_vals $\leftarrow $ [0, 2pi/intervals, 2* 2pi/intervals, .....]     
$\Delta \leftarrow 2pi/intervals$  

from i in range (0, intervals)    
&emsp; left $\leftarrow$ (kidney_eq(interval_vals[i]))^2      
&emsp; right $\leftarrow$ (kidney_eq(interval_vals[i+1]))^2     
&emsp; area += ((left + right)/2) * $\Delta$ * 0.5     

return area - disk_area 
#### Code

In [10]:
import numpy as np 

def disk_area(r_squared):  
    # We are given r^2 in the formuala for the disk so let the user pass it in 
    return r_squared * np.pi 

def kidney_equation(x):  
    # Equation of Kidney after translating to Polar, x is the angle plugged in!
    return np.cos(x) ** 3 + np.sin(x) ** 3 

def trapezoidal_method_polar(kidney_eq, intervals, dug_area):  
    """ 
    Function For Trapezoidal Method on Polar Equation with cut out area! 

    Input: 
    - kidney_eq: equation for the provided kidney (Could be generalized to function of curve) 
    - intervals: Whole number of intervals to be used in calculating 
    - dug_area: Area of disk that's been cut out of the kidney 

    Output: 
    - result: The approximation of the area under the curve after subtracting out the dug area 
              using trapezoidal method
    """ 
    interval_vals = np.linspace(0, 2*np.pi, intervals+1) 
    step = (2*np.pi)/intervals  
    area = 0
    for i in range(len(interval_vals)-1):    
        left_value = (kidney_eq(interval_vals[i])) ** 2
        right_value = (kidney_eq(interval_vals[i+1])) **2
        area += 0.5 * step * ((left_value + right_value)/2)
    return area-dug_area

dug_area = disk_area(0.125) 

ans = trapezoidal_method_polar(kidney_equation, 10000, dug_area) 
print(f"Approximated Area: {ans}") 

Approximated Area: 1.5707963267949026


### Results 

The results of my program were the following: 

Area of Kidney with Cut Disk $\approx$ 1.570796 

### Performance 

In terms of the performance of this trapezoidal method, the time complexity would be O(n) where n   
represents the number of intervals we decided to use as that's the number of computations performed.  
The space complexity of the program is also O(n) as we have to store all of these intervals. There is     
not much else to comment on this program's performance besides it being a good way to approximate the    
area under a curve as computing integrals is computationally expensive. 

## Background For Problem 2.2 

<div style="text-align: center;">
    <img src="Problem_2.2_Background.png" alt="Description" style="width:1000px; border:2px solid black;"/>
</div> 



## Problem 2.2

### Problem Description 

Given some N x N matrix called A filled with uniformly distributed floating-point random numbers,   
and a vector, b, of size n with 1 as all of it's elements, write a program to solve AX = b, if N = 66.   

### Algorithm Description  


#### Psuedocode 

#### Code 

In [ ]:
""" 
We have to somehow randomly create a matrix of uniformly distributed floating point values of size 63. 
Need to generate N by N values so 63 * 63 values of unifromly distributed floating point numbers 
Generating B of size N is not hard, it's just the matrix of 1's, transposed  

How do we solve this numerically?  
- Gaussian Elimination?  
    Forward Elimination and Then Back Substitution! 
    Forward Elimination Algorithm would be find non-zero pivot. Make it the first row for that column we're pivoting in 
        IE: 1st Column = 1st Row, 2nd = 2nd row and so forth 
        After doing this swap, we scale and subtract it from everything below to make it 0    
        At the end we get everything with ones and other values 
    Backward Elimination
        We take the last value, and we subtract it from everything 
    
    To get started let's figure out how to set up the problem using the numpy library 
    np.random.uniform(low=, high=, size=(m, n)) sets up a numpy array of this size, giving us our matrix! 
    np.ones((m, n)) generates all intialized values to one for a matrix of m * n 
    np.hstack((A, b))

"""
import numpy as np  

def gaussian_elimination(N): 
    """
    Inputs: 
    - N: Size of N * N matrix to be randomly generated 
    Outputs: 
    - x: A vector of size N that is the solution to Ax = b 
    
    """
    A = np.random.uniform(low=-1, high=1, size=(N, N))  
    b = np.ones((N,1)) 
    Ab = np.hstack((A,b)) 
    # print(A) 
    # print(b) 
    # print(Ab)
    for i in range(N):  
        max_val = 0 
        max_row = i
        for j in range(i, N):  
            if(abs(Ab[j][i])) > max_val: 
                max_val = abs(Ab[j][i]) 
                max_row = j 
        if i != max_row:   
            # Numpy Row Swapping Below! 
            Ab[[i, max_row]] = Ab[[max_row, i]] 

        # Now we can actually start eliminating below our max row which should be i + 1 
        for k in range(i+1, N): 
            factor = Ab[k, i]/Ab[i, i] 
            Ab[k, i:] -= factor * Ab[i, i:]   
    # print(Ab) 
    # print("Forward Elimination Done")
    # Now Ab should be done, so we backsubstitute to solve our equation 
    x = np.zeros(N) 
    for i in range(N-1, -1, -1):  
        # b - all other x_values is what we need to be done  
        non_target_x_vals = np.sum(Ab[i, i+1:N] * x[i+1 : N])
        x[i] = (Ab[i, -1] - non_target_x_vals)/(Ab[i, i])  
    
    x_np = np.linalg.solve(A,b)  
    print(f"Answer from Numpy's to Double Check:")  
    print(x_np.flatten()) 
    print(f"\n")
    print(f"Answer from our implementation:\n") 
    print(x)
    return x 

x = gaussian_elimination(66)  
print("")


Forward Elimination Done
Answer from Numpy's to Double Check:
[ 1.63782768  3.60201161  1.6560252  -3.10503699  2.19104336 -0.12246754
 -1.35453016  2.086222   -2.47532635  1.3337225  -1.31498671  1.47891578
  0.29312559  0.70053039 -0.15695219 -2.50219664 -1.80509211  0.39896173
 -0.51476741  2.54059047 -0.2045843  -0.54103374  1.87346736  2.84282363
  0.43623375 -1.05154721 -0.08224175  0.82615802 -1.22130369  1.78543955
 -3.15101789 -1.67967581 -0.60770558  0.65254782 -0.34525938 -1.32000886
 -5.32597467  0.96789325  0.45882931  0.02526064  1.34324628  1.03478795
 -0.61005817  0.90305578 -2.05430685  1.85068074  0.71922019 -3.33909377
 -0.24930608 -0.23552784 -0.3564455   0.4781699  -1.0206195   1.01830744
  0.68341477  1.44163815  0.35442709 -1.6772874  -0.97339785  0.5371998
 -0.51747917 -1.91734343 -4.17774943 -2.12281387 -0.37920467 -0.43913761]


Answer from our implementation:

[ 1.63782768  3.60201161  1.6560252  -3.10503699  2.19104336 -0.12246754
 -1.35453016  2.086222   -2